web3/middleware/async_middleware/async_cache.py

```diff
@@ -0,0 +1,104 @@
+import functools
+import threading
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    Collection,
+    Dict,
+    Tuple,
+    Type,
+    cast,
+)
+
+import lru
+
+from web3._utils.caching import (
+    generate_cache_key,
+)
+from web3.types import (
+    AsyncMiddleware,
+    Middleware,
+    RPCEndpoint,
+    RPCResponse,
+)
+
+if TYPE_CHECKING:
+    from web3 import Web3  # noqa: F401
+
+
+SIMPLE_CACHE_RPC_WHITELIST = cast(
+    Tuple[RPCEndpoint],
+    (
+        "web3_clientVersion",
+        "eth_getBlockTransactionCountByHash",
+        "eth_getUncleCountByBlockHash",
+        "eth_getBlockByHash",
+        "eth_getTransactionByHash",
+        "eth_getTransactionByBlockHashAndIndex",
+        "eth_getRawTransactionByHash",
+        "eth_getUncleByBlockHashAndIndex",
+        "eth_chainId",
+    ),
+)
+
+
+def _should_cache_response(
+    _method: RPCEndpoint, _params: Any, response: RPCResponse
+) -> bool:
+    return (
+        "error" not in response
+        and "result" in response
+        and response["result"] is not None
+    )
+
+
+async def async_construct_simple_cache_middleware(
+    cache_class: Type[Dict[Any, Any]],
+    rpc_whitelist: Collection[RPCEndpoint] = SIMPLE_CACHE_RPC_WHITELIST,
+    should_cache_fn: Callable[
+        [RPCEndpoint, Any, RPCResponse], bool
+    ] = _should_cache_response,
+) -> Middleware:
+    """
+    Constructs a middleware which caches responses based on the request
+    ``method`` and ``params``
+
+    :param cache_class: Any dictionary-like object
+    :param rpc_whitelist: A set of RPC methods which may have their responses cached.
+    :param should_cache_fn: A callable which accepts ``method`` ``params`` and
+        ``response`` and returns a boolean as to whether the response should be
+        cached.
+    """
+
+    async def async_simple_cache_middleware(
+        make_request: Callable[[RPCEndpoint, Any], Any], async_w3: "Web3"
+    ) -> AsyncMiddleware:
+        cache = cache_class()
+        lock = threading.Lock()
+
+        async def middleware(method: RPCEndpoint, params: Any) -> RPCResponse:
+            if method in rpc_whitelist:
+                with lock:
```

Comment:
Interested on feedback in this main section of the thread locking

Reply:
This is the exact issue I currently have in request.py with async and thread locking. The below await will release the execution of this method for something else to work while you are still in the lock. That will eventually cause a deadlock.

Reply:
Yeah I think I get it a bit more now. Thanks for the explanation. I think the first step here is probably writing a good test that fails, with the situation you described, and work to make it not fail. I'll see if I can whip something up this week but I'm out for half of the week. Realistically probably won't get to this until next week.

Reply:
I was able to use the POC in #2446 to pretty reliably reproduce the request.py issue, so maybe you could use something similar.